src/cudamatrix/cu-sparse-matrix.h

```diff
@@ -140,79 +122,123 @@ class CuSparseMatrix {
+                     kNoTrans) :
+      num_rows_(0), num_cols_(0), nnz_(0), csr_row_ptr_col_idx_(NULL), csr_val_(
+      NULL) {
     this->CopyFromSmat(smat, trans);
   }
 
-
   /// Constructor from an array of indexes.
   /// If trans == kNoTrans, construct a sparse matrix
   /// with num-rows == indexes.Dim() and num-cols = 'dim'.
   /// 'indexes' is expected to contain elements in the
-  /// range [-1, dim - 1].  Each row 'i' of *this after
-  /// calling the constructor will contain zero elements
-  /// if indexes[i] == -1, otherwise it will contain a single
+  /// range [0, dim - 1].  Each row 'i' of *this after
+  /// calling the constructor will contain  a single
   /// element at column-index indexes[i] with value 1.0.
   ///
   /// If trans == kTrans, the result will be the transpose
   /// of the sparse matrix described above.
-  ///  TODO: implement this.
-  CuSparseMatrix(const CuArray<int32> &indexes,
-                 int32 dim, MatrixTransposeType trans = kNoTrans);
+  CuSparseMatrix(const CuArray<int32> &indexes, int32 dim,
+                 MatrixTransposeType trans = kNoTrans);
 
   /// Constructor from an array of indexes and an array of
   /// weights; requires indexes.Dim() == weights.Dim().
   /// If trans == kNoTrans, construct a sparse matrix
   /// with num-rows == indexes.Dim() and num-cols = 'dim'.
   /// 'indexes' is expected to contain elements in the
-  /// range [-1, dim - 1].  Each row 'i' of *this after
-  /// calling the constructor will contain zero elements
-  /// if indexes[i] == -1, otherwise it will contain a single
+  /// range [0, dim - 1].  Each row 'i' of *this after
+  /// calling the constructor will contain a single
   /// element at column-index indexes[i] with value weights[i].
   /// If trans == kTrans, the result will be the transpose
   /// of the sparse matrix described above.
-  ///  TODO: implement this.
   CuSparseMatrix(const CuArray<int32> &indexes,
-                 const CuVector<Real> &weights,
-                 int32 dim, MatrixTransposeType trans = kNoTrans);
+                 const CuVectorBase<Real> &weights, int32 dim,
+                 MatrixTransposeType trans = kNoTrans);
 
-  ~CuSparseMatrix() { }
+  ~CuSparseMatrix() {
+    Destroy();
+  }
 
- protected:
+protected:
   // The following two functions should only be called if we did not compile
   // with CUDA or could not get a CUDA card; in that case the contents are
   // interpreted the same as a regular sparse matrix.
   inline const SparseMatrix<Real> &Mat() const {
-    return *(reinterpret_cast<const SparseMatrix<Real>* >(this));
+    return *(reinterpret_cast<const SparseMatrix<Real>*>(this));
   }
   inline SparseMatrix<Real> &Mat() {
-    return *(reinterpret_cast<SparseMatrix<Real>* >(this));
+    return *(reinterpret_cast<SparseMatrix<Real>*>(this));
+  }
+
+  /// Returns pointer to the data array of lenght nnz_ that holds all nonzero
+  /// values in zero-based CSR format
+  const Real* CsrVal() const {
```

Comment:
It's not clear to me that it's the right decision, to make these things public.  Is it feasible to make these things private and make use of friend declarations?  I want to leave the door open to refactoring this class later on if there is a need.